corehq/motech/fhir/tests/test_matchers.py

```diff
@@ -0,0 +1,136 @@
+from decimal import Decimal
+from uuid import uuid4
+
+from django.test import SimpleTestCase
+
+from nose.tools import assert_equal
+
+from ..const import SYSTEM_URI_CASE_ID
+from ..matchers import (
+    GivenName,
+    NegativeIdentifier,
+    OrganizationMatcher,
+    OrganizationName,
+    PatientMatcher,
+    PersonMatcher,
+    PropertyWeight,
+)
+
+
+def test_given_name_method():
+    """
+    Test the GivenName ComparisonMethod as used by PersonMatcher
+    """
+    pw = PropertyWeight('$.name[0].given', Decimal('0.3'), GivenName)
+    assert pw in PersonMatcher.property_weights
+    method = pw.method
+
+    john = {'name': [{'given': 'Henry John Colchester'.split()}]}
+    for resource, candidate, expected in [
+        # candidate "H. John" matches patient "Henry John Colchester"
+        (john, {'name': [{'given': ['H.', 'John']}]}, True),
+        # candidate "John" does not match patient "Henry John Colchester"
+        (john, {'name': [{'given': ['John']}]}, False),
+        (john, {'name': [{'given': ['Henry']}]}, True),
+        (john, {'name': [{'given': ['H.', 'J.', 'C.']}]}, False),
+        # candidate "Eric John Marwood" matches "Henry John Colchester"
+        (john, {'name': [{'given': ['Eric', 'John', 'Marwood']}]}, True),
```

Comment:
Yeah, but is that desirable? My child goes by her middle name, so this wouldn't find her case, right?

Reply:
Exactly! "Henry John Colchester" is my grandfather. The variable is `john` because he went by his second name. So this test (and the last example especially) is meant to illustrate the strengths and weaknesses of this approach.

We could also define an `AnyGivenName` like ...
```python
class AnyGivenName(ComparisonMethod):

    @staticmethod
    def compare(a, b):
        return bool(set(a).intersection(set(b)))
```
... but it it would need a very small weight, because it would match ['H.', 'John'] and ['Santa', 'H.'].

In the end I left it out, because I wasn't sure it was as useful as it first seemed. I've tried to be guided by "How do we avoid false positives?" rather than by "How do we match patients?" I think there is a lot more untapped potential in negative weights.